### PR TITLE
`<datetime expression> = <date literal string>` should do something sane

### LINKED ISSUE #17807: Adding an "on" date filter without a specific time to a custom date column returns no results
**Describe the bug**
Adding an "on" date filter without a time to a custom date column shows no results. Filtering on the exact date value works, and filtering using the "between" option works, but you can't filter on a date like `10/07/2019` and see rows with values like `10/07/2019 01:34am`

**Logs**
Here's an example query we send to the `/api/dataset` endpoint. See repro steps below.
```
{"database":1,"query":{"source-table":3,"expressions":{"CC Birth Date":["field",24,null],"CC Created At":["field",18,null]},"fields":[["field",30,null],["field",21,null],["field",23,null],["field",19,null],["field",22,null],["expression","CC Created At"],["field",29,null],["field",20,null],["field",25,null],["field",26,null],["field",24,{"temporal-unit":"default"}],["field",28,null],["field",27,null],["field",18,{"temporal-unit":"default"}],["expression","CC Birth Date"]],"filter":["=",["expression","CC Created At"],"2017-10-07"]},"type":"query","parameters":[]}
```

**To Reproduce**
1. Ask a Question > Custom Question > People dataset
2. Add a custom column that uses the expression `[Created At]`. Name it `CC Created At`
Screen Shot 2021-09-08 at 11 17 58 AM

3. Run the query. Click on a cell in the "CC Created At" column and select the `=` filter. You should see a filter like "Created At is October 7, 2017 01:34 am" above the table and the table should contain a single result row.

Screen Shot 2021-09-08 at 11 16 12 AM

4. Click on the filter to modify it. Add the same date to it but without a time. Run the query and observe that there are now no results.

Screen Shot 2021-09-08 at 11 16 26 AM

5. You can do this using the original "Created At" field and it'll work correctly.

**Expected behavior**
You'd expect custom column date fields to filter identically to how normal date fields filter.

**Information about your Metabase Installation:**
This is on master, HEAD at `0b60ff9234797cb465dbf4040e2d180120b2fbc1`

**Severity**
P2

### PR DESCRIPTION
**Pull request by camsaul**
Fixes https://github.com/metabase/metabase/issues/17807

For a filter clause like 

```clj
[:= {} [:expression "expr" {:effective-type :type/DateTime}] "2023-11-08"]
```

we were actually generating SQL like

```clj
WHERE expr = ? -- #t "2023-11-08T00:00:00-00:00"
```

which while valid at least is clearly not what people meant. This PR tweaks filter optimization logic to detect situations like this and instead output "optimized" date range filters like

```sql
WHERE expr >= ? AND expr < ?-- #t "2023-11-08T00:00:00-00:00", #t "2023-11-09T00:00:00-00:00"
```

The original version of this PR had a lot more changes and fancy optimization stuff, I ended up removing that in the end since it was too hard to reconcile it with other fixes that have happened since November, so this final version just has the parts needed to fix #17807

### CODE REVIEW

File: src/metabase/driver.clj [hunk 1]
```diff
@@ -31,6 +32,8 @@
   `nil` (meaning subsequent queries will not attempt to change the session timezone) or something considered invalid
   by a given Database (meaning subsequent queries will fail to change the session timezone)."
   []
+  (when config/tests-available?
+    ((requiring-resolve 'mb.hawk.parallel/assert-test-is-not-parallel) `notify-all-databases-updated))
```
Comment (camsaul) [inline src/metabase/driver.clj:36]: We should not be doing this in parallel tests, for Snowflake it kills the open connection pool which will break other tests running at the same time.

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 2]
```diff
@@ -104,75 +115,253 @@
...
+  (case (first literal)
+    :absolute-datetime
+    (let [[_absolute-datetime t unit] literal]
+      (when (= unit interval-unit)
+        [:absolute-datetime (u.date/add t unit interval-n) unit]))
+
+    :relative-datetime
+    (let [[_relative-datetime n unit] literal]
+      (if (= n :current)
+        [:relative-datetime interval-n interval-unit]
```
Comment (bshepherdson) [inline src/metabase/query_processor/middleware/optimize_temporal_filters.clj:330]: I can't convince myself that this is sound in all cases, but maybe I'm thinking of it wrong.

My vague mental model is that a `:relative-datetime` defines a window of time (say a month) and the interval gives a delta like `+ 3 weeks`. In my mind adding 3 weeks to "this month" turns `Nov 1 - Nov 30, 2023` into `Nov 22 - Dec 21, 2023`.

But if I'm understanding how `:relative-datetime` works properly, this code is dropping the original window size (month) and replacing it with "3 weeks from this week". So for example, starting from Friday Nov 24 and using Monday as the first day, that rounds up to the next week (Mon Nov 27) and then adds two more, resulting in Mon Dec 11 to Sun Dec 17.

Do we want it to translate a same-sized window, or is changing the window intended? Or is there a missing `(= unit interval-unit)` check for this case?

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 3]
```diff
@@ -104,75 +115,253 @@
...
+       (let [{:keys [non-interval intervals]} (group-temporal-arithmetic-args expr)
+             intervals                        (optimize-intervals intervals)]
+         (when (and (mbql.u/is-clause? #{:absolute-datetime :relative-datetime} non-interval)
+                    (= (count intervals) 1))
+               (let [interval (first intervals)
+                     interval (case tag
+                                :+ interval
+                                :- (let [[_interval n unit] interval]
+                                     [:interval (- n) unit]))]
+             (add-interval-if-possible non-interval interval)))))
```
Comment (bshepherdson) [inline src/metabase/query_processor/middleware/optimize_temporal_filters.clj:353]: nit: weird indentation.

File: test/metabase/driver/sql/query_processor_test.clj [hunk 4]
```diff
@@ -254,22 +255,26 @@
...
+                      [:cast
+                       [:week (h2x/with-database-type-info
+                               [:dateadd (h2x/literal "day")
+                                [:inline 1]
+                                (h2x/with-database-type-info [:cast :created_at [:raw "datetime"]] "datetime")]
+                               "datetime")]
+                       [:raw "datetime"]]
+                      "datetime")]
+                    "datetime"))
+               (sql.qp/adjust-start-of-week :h2 (fn [x] [:week x]) :created_at)))))
```
Comment (bshepherdson) [inline test/metabase/driver/sql/query_processor_test.clj:272]: Optional style/taste thing: I find `(f some-big-long-expression short-value)` hard to read, and
`(-> some-big-long-expression (f short-value))` easier.

File: test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj [hunk 5]
```diff
@@ -10,21 +10,26 @@
    [metabase.test :as mt]
    [metabase.util.date-2 :as u.date]))
 
-(driver/register! ::timezone-driver, :abstract? true)
+(deftest ^:parallel temporal-arithmetic?-test
+  (is (#'optimize-temporal-filters/temporal-arithmetic?
+       [:+ [:field 1 {:base-type :type/DateTime}] [:interval -2 :month]])))
```
Comment (bshepherdson) [inline test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj:15]: A couple of negative tests wouldn't be a bad idea.

**Review by bshepherdson — APPROVED** [review]

File: test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj [hunk 6]
```diff
@@ -10,21 +10,26 @@
...
 
-(defmethod driver/database-supports? [::timezone-driver :set-timezone] [_driver _feature _db] true)
+(deftest ^:parallel group-temporal-arithmetic-args-test
+  (is (= {:tag          :+
+          :intervals    [[:interval -2 :month]]
+          :non-interval [:field 1 {:base-type :type/DateTime}]}
+         (#'optimize-temporal-filters/group-temporal-arithmetic-args
+          [:+ [:field 1 {:base-type :type/DateTime}] [:interval -2 :month]]))))
```
Comment (metamben) [inline test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj:22]: Tests for multiple values and no value could be interesting.

**Review by metamben — APPROVED** [review]
Really nice! 🥇 I found the metadata-provider changes in this PR a bit distracting. I'm also concerned by the different unit thing Braden pointed out, but stamping anyway.

File: src/metabase/mbql/schema.cljc [hunk 7]
```diff
@@ -554,14 +555,15 @@
 (def ^:private ExpressionArg
   [:ref ::ExpressionArg])
 
-(mr/def ::NumericExpressionArgOrInterval
+(mr/def ::Addable
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/mbql/schema.cljc:558]: That's totally true. But the old name was wrong (it was actually a DateTimeExpression, NumericExpression, or an interval). What name makes more sense here?

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: fixed

File: src/metabase/mbql/schema.cljc (hunk 7 above)
[the remark this answers is not in the document]
Comment (metamben) [reply src/metabase/mbql/schema.cljc:558]: Dunno, nothing better than spelling out the full name comes to my mind. Or can we name the context in which it occurs somehow?

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: ended up removing this stuff

File: test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj (hunk 5 above)
Comment (camsaul) [reply test/metabase/query_processor/middleware/optimize_temporal_filters_test.clj]: ended up removing this stuff

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 8]
```diff
@@ -2,66 +2,77 @@
...
+
+(defn- interval? [x]
+  (mbql.u/is-clause? :interval x))
+
+(defn- temporal-arithmetic? [x]
+  (and (mbql.u/is-clause? #{:+ :-} x)
+       (let [[_tag & args] x]
+         (and (every? (partial mbql.u/is-clause? #{:relative-datetime :absolute-datetime :interval :field :expression})
+                      args)
+              (= (count (filter (some-fn temporal-literal? temporal-ref?) args)) 1)))))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: ended up removing this stuff

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 9]
```diff
@@ -104,75 +115,253 @@
...
+
+(mu/defn ^:private group-temporal-arithmetic-args :- [:map
+                                                      [:tag          [:enum :+ :-]]
+                                                      [:non-interval [:or
+                                                                      mbql.s/field
+                                                                      mbql.s/expression
+                                                                      mbql.s/relative-datetime
+                                                                      mbql.s/absolute-datetime]]
+                                                      [:intervals    [:maybe [:sequential mbql.s/interval]]]]
+  [[tag & args :as _expr] :- [:or mbql.s/+ mbql.s/-]]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: ended up removing this stuff

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 10]
```diff
@@ -104,75 +115,253 @@
...
+          [::arithmetic-wrapping-ref ::literal ::literal]
+          (let [{x' :x, lower' :y} (transfer-temporal-arithmetic-args x lower)
+                {upper' :y}        (transfer-temporal-arithmetic-args x upper)]
+            (optimize [:between x' lower' upper']))
+          #_else
+          nil)
+        [:between x lower upper])))
+
+(mu/defn ^:private combine-intervals-if-possible :- [:maybe mbql.s/interval]
+  "Combine two `:interval` clauses into one if they both have the same unit."
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: ended up removing this stuff

File: src/metabase/query_processor/middleware/optimize_temporal_filters.clj [hunk 11]
```diff
@@ -104,75 +115,253 @@
...
+  "Combine multiple `:interval` clauses into the fewest number possible."
+  [intervals :- [:sequential {:min 1} mbql.s/interval]]
+  (let [intervals (sort-by
+                   (fn [[_interval _n unit]]
+                     unit)
+                   intervals)]
+    (reduce
+     (fn [acc interval]
+       (if (empty? acc)
+         [interval]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/optimize_temporal_filters.clj]: ended up removing this stuff

File: test/metabase/driver/sql/query_processor_test.clj (hunk 4 above)
Comment (camsaul) [reply test/metabase/driver/sql/query_processor_test.clj]: fixed

**Pull request merged**